federator/controllers/amkocluster_controller.go

```diff
@@ -51,190 +52,246 @@ type AMKOClusterReconciler struct {
 	if err != nil {
-		statusErr := r.UpdateAMKOClusterStatus(ctx, FederationTypeStatus, StatusMsgFederationFailure, err.Error(), &amkoCluster)
-		if statusErr != nil {
-			return ctrl.Result{
-				RequeueAfter: time.Second * 5,
-			}, statusErr
-		}
-		return reconcile.Result{}, err
+		return ctrlResultRequeue, err
 	}
 
-	// fetch the member clusters' contexts
-	memberClusters, err := FetchMemberClusterContexts(ctx, amkoCluster.DeepCopy())
+	// validate member clusters and update status
+	validClusters, err := r.ValidateMemberClustersAndUpdateStatus(ctx, memberClusters, updatedAMKOCluster)
 	if err != nil {
-		statusErr := r.UpdateAMKOClusterStatus(ctx, FederationTypeStatus, StatusMsgFederationFailure, err.Error(), &amkoCluster)
-		if statusErr != nil {
-			return ctrl.Result{
-				RequeueAfter: time.Second * 5,
-			}, statusErr
-		}
-		return reconcile.Result{}, fmt.Errorf("error in fetching member cluster contexts: %v", err)
+		return ctrlResultRequeue, err
+	}
+
+	// Federate the GSLBConfig object on all member clusters
+	if err := r.FederateGSLBConfigAndUpdateStatus(ctx, validClusters, updatedAMKOCluster); err != nil {
+		return ctrlResultRequeue, err
 	}
 
-	// validate member clusters
-	err = ValidateMemberClusters(ctx, memberClusters, amkoCluster.Spec.Version)
+	// Federate the GDP object on all member clusters
+	if err := r.FederateGDPAndUpdateStatus(ctx, validClusters, updatedAMKOCluster); err != nil {
+		return ctrlResultRequeue, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *AMKOClusterReconciler) FetchMemberClusterContextsAndUpdateStatus(ctx context.Context, amkoCluster *amkov1alpha1.AMKOCluster) ([]KubeContextDetails, error) {
+	memberClusters, errClusters, err := FetchMemberClusterContexts(ctx, amkoCluster.DeepCopy())
+
 	if err != nil {
-		statusErr := r.UpdateAMKOClusterStatus(ctx, FederationTypeStatus, StatusMsgFederationFailure, err.Error(), &amkoCluster)
-		if statusErr != nil {
-			return ctrl.Result{
-				RequeueAfter: time.Second * 5,
-			}, statusErr
+		// update the error message in the AMKOCluster status field
+		if statusErr := r.UpdateAMKOClusterStatus(ctx, ClusterContextsStatusType, "", err.Error(),
+			errClusters, amkoCluster); statusErr != nil {
+			return nil, statusErr
+		}
+		// errors on which the execution should be stopped here and retried:
+		// - bad kubeconfig
+		// - error in generating a temporary kubeconfig file
+		return nil, fmt.Errorf("error in fetching member cluster contexts: %v", err)
+	} else {
+		if statusErr := r.UpdateAMKOClusterStatus(ctx, ClusterContextsStatusType, "", "",
+			errClusters, amkoCluster); statusErr != nil {
+			return nil, statusErr
 		}
-		return reconcile.Result{}, fmt.Errorf("error in validating the member clusters: %v", err)
 	}
 
-	// Federate the GSLBConfig object on all member clusters
-	err = r.FederateGSLBConfig(ctx, memberClusters)
+	// if memberClusters list is empty, clear the rest of the status fields and return
```

Comment:
Looks are status is not getting cleaned up, although the comment says so.

Reply:
removed this comment. status is not required to be cleaned up.